blocks/bricks/sequence_generators.py

```diff
@@ -20,49 +49,72 @@ class BaseSequenceGenerator(Initializable):
 
     This class combines two components, a readout network and an
     attention-equipped recurrent transition, into a context-dependent
-    sequence generator. Optionally a third component can be used which
+    sequence generator. Third component must be also given which
     forks feedback from the readout network to obtain inputs for the
     transition.
 
-    **Definitions:**
+    The class provides two methods: :meth:`generate` and :meth:`cost`. The
+    former is to actually generate sequences and the latter is to compute
+    the cost of generating given sequences.
 
-    * *States* of the generator are the states of the transition as
-      specified in `transition.state_names`.
+    The generation algorithm description follows.
 
-    * *Contexts* of the generator are the contexts of the transition as
-      specified in `transition.context_names`.
+    **Definitions and notation:**
 
-    * *Glimpses* are intermediate entities computed at every generation
-      step from states, contexts and the previous step glimpses. They are
-      computed in the transition's `apply` method when not given or by
-      explicitly calling the transition's `take_glimpses` method. The set
-      of glimpses considered is specified in `transition.glimpse_names`.
+    * States :math:`s_i` of the generator are the states of the transition
+      as specified in `transition.state_names`.
 
-    The generation algorithm description follows.
+    * Contexts of the generator are the contexts of the
+      transition as specified in `transition.context_names`.
+
+    * Glimpses :math:`g_i` are intermediate entities computed at every
+      generation step from states, contexts and the previous step glimpses.
+      They are computed in the transition's `apply` method when not given
+      or by explicitly calling the transition's `take_glimpses` method. The
+      set of glimpses considered is specified in
+      `transition.glimpse_names`.
+
+    * Outputs :math:`y_i` are produced at every step and form the output
+      sequence. A generation cost :math:`c_i` is assigned to each output.
 
     **Algorithm:**
 
-    1. The initial states are computed from the contexts. This includes
-       fake initial outputs given by the `initial_outputs` method of the
-       readout, initial states and glimpses given by the `initial_state`
-       method of the transition.
+    1. Initialization.
+
+       :math:`y_0 = readout.initial\_outputs(contexts)`
+
+       :math:`s_0, g_0 = transition.initial\_states(contexts)`
```

Comment:
Right, `s_0` does not depend on anything, it is determined by the transition itself. Options included fully zero state, a trainable one, or one predicted from the contexts. 

The reason why we have `y_0` and `f_0` is that there exists a popular hack to use `y_{i-1}`, more specifically `f_{i-1}`, to predict `y_i`. By default this does not happen though, one should include 'feedback' into the `source_names` of the readout to switch it on. This is why `y_0` is needed, to be used for prediction of `y_1`.

Your edit sounds like a nice summary for why we need `y_0` :)